scripts-base/src/Component.ts

```diff
@@ -0,0 +1,96 @@
+import {
+	ComponentEl,
+	DelegateEvent,
+	DelegateTarget,
+	EventListeners,
+	EventMap,
+	EventMapByElement,
+} from './componentTypes'
+import { matchesSelector } from './utils'
+
+export interface NamedComponent {
+	componentName: string
+}
+
+export type ComponentConstructor<
```

Comment:
Není pojmenování `ComponentConstructor` zavádějící, když se používá pouze pro `initializeComponents.components`, ale ne v samotné `Component` třídě?

Reply:
Jak je tady místo použití relevantní? Je to _component constructor_, tak se to jmenuje `ComponentConstructor`. To jméno mi přijde naprosto v pořádku. Jak jinak bys to vůbec pojmenoval?

Reply:
`constructor` -> `constructor()` což tady neplatí. V tomhle případě mi to přijde jako ~ `initializer` (od toho i název `initComponents`)

Reply:
Platí. `constructor()` je class-based syntax sugar pro constructor function (normální JS funkci, kterou lze zavolat s `new`), což je přesně to, co se tady typuje. Konstruktor, kterej vrací instanci `Component`.

Reply:
Narážím na:

[Component `constructor`:](https://github.com/manGoweb/npm/blob/6c62e775f9d05c140a561f3ab566bef8aba4842d/scripts-base/src/Component.ts#L30)
```typescript
constructor(protected readonly el: ComponentElement, protected readonly data: Data) {}
```

[Type `ComponentConstructor`](https://github.com/manGoweb/npm/blob/6c62e775f9d05c140a561f3ab566bef8aba4842d/scripts-base/src/Component.ts#L15):
```typescript
type ComponentConstructor<
	Data,
	ComponentElement extends ComponentEl,
	EMap ...>
```

[Použití typu `ComponentConstructor`](https://github.com/manGoweb/npm/blob/b25e28a30cc2d08282d0aa4ea77fba0afe02f850/scripts-base/src/initializeComponents.ts#L3-L34):
```typescript
const initializeComponents = (
	components: Array<ComponentConstructor<any, any, any>>,
	...
	componentsByName[name] = components[i]
	...
	const Component = componentsByName[component.name] // class
	...
	const instance = new Component(placement, component.data || {})
```

Kde se neshoduje function signature:
Konstruktor `(element, data)` vs ComponentConstructor `(data, selector || element, events)`

Reply:
Ty pleteš separátní věci dohromady. To, že ten **typ** přijímá jiné **typové parametry**, nemá absolutně co dočinění s tím, jaké parametry přijímá ta **hodnota**. To `new Component(placement, component.data || {})` úplně přesně odpovídá tomu `constructor(protected readonly el: ComponentElement, protected readonly data: Data)`. Ten `ComponentConstructor` potřebuje tři **typové parametry**, abychom dokázali vykouzlit, jaký typ budou mít ty parametry, které přijímá ta _constructor function_.